portal/src/main/webapp/js/src/study-view/data/StudyViewProxy.js

```diff
@@ -412,6 +413,117 @@ var StudyViewProxy = (function() {
         return deferred.promise();
     }
 
+    //Change the CNA data base on sample ID
+    function convertCNAData(data) {
+        var converted = {};
+
+        if (data) {
+            for (var i = 0, dataL = data.gene.length; i < dataL; i++) {
+                var caseIds = data.caseIds[i];
+
+                for (var j = 0, caseIdsL = caseIds.length; j < caseIdsL; j++) {
+                    if (!converted.hasOwnProperty(caseIds[j])) {
+                        converted[caseIds[j]] = [];
+                    }
+                    converted[caseIds[j]].push({
+                        alter: data.alter[i],
+                        cytoband: data.cytoband[i],
+                        gene: data.gene[i],
+                        gistic: data.gistic[i]
+                    });
+                }
+            }
+        }
+        return converted;
+    }
+
+    function getCNABasedOnSampleIds(sampleIds) {
+        var data = {
+            alter: [],
+            caseIds: [],
+            cytoband: [],
+            gistic: [],
+            gene: []
+        };
+        if (sampleIds instanceof Array) {
+            if (sampleIds.length === this.getSampleIds().length) {
+                return obtainDataObject.cna;
+            }
+            if (obtainDataObject.hasOwnProperty('cnaSampleBased')) {
+                var geneSpecific = {};
+                var numOfSample = sampleIds.length;
+                for (var i = 0; i < numOfSample; i++) {
+                    if (obtainDataObject.cnaSampleBased.hasOwnProperty(sampleIds[i])) {
+                        for (var j = 0, numOfGenes = obtainDataObject.cnaSampleBased[sampleIds[i]].length; j < numOfGenes; j++) {
+                            var key = obtainDataObject.cnaSampleBased[sampleIds[i]][j].gene + obtainDataObject.cnaSampleBased[sampleIds[i]][j].alter;
+                            if (!geneSpecific.hasOwnProperty(key)) {
+                                geneSpecific[key] = obtainDataObject.cnaSampleBased[sampleIds[i]][j];
+                                geneSpecific[key].caseIds = [];
+                            }
+                            geneSpecific[key].caseIds.push(sampleIds[i]);
+                        }
+                    }
+                }
+                _.values(geneSpecific).map(function (item) {
```

Comment:
this loops again over the geneSpecific list. It is better (and also faster) to do this immediately after line 460, right? So just after line 460 above, add:
```javascript
                   data.alter.push(geneSpecific[key].alter);
                   data.caseIds.push(geneSpecific[key].caseIds);
                   data.cytoband.push(geneSpecific[key].cytoband);
                   data.gistic.push(geneSpecific[key].gistic);
                   data.gene.push(geneSpecific[key].gene);
```